### PR TITLE
Core: Addition of a Hook in CanFlyInZone check

### PR DESCRIPTION
**Pull request by Tralenor**
## Changes Proposed:
-  Add a bool Hook in CanFlyInZone check to give modules the possibility to influence this function to deny flight based on modules decision.
-  

## Tests Performed:

- Builds without errors, tested ingame without any modules -> has no impact as intended(aka no regression for servers without module), tested with Proof-of-Concept module: [poc-can-fly-hook-override](https://github.com/Tralenor/poc-can-fly-hook-override), has intended impact, disallowing flying in Hellfire Peninsula. 

## How to Test the Changes:

you can use the Module:  [poc-can-fly-hook-override](https://github.com/Tralenor/poc-can-fly-hook-override) to test this hook, this module simply blocks flight over the Hellfire Peninsula.

Test scenario 0 (check for regression, without having the module):
0. checkout revision and compile
1. go ingame, try to fly with your prefered fly-mount in various locations, eg. Loch Modan, Borean Tundra and outland and validate that it behaves blizzlike.

Test scenario 1:
0. add module [poc-can-fly-hook-override](https://github.com/Tralenor/poc-can-fly-hook-override) to modules subfolders.
1. checkout revision and compile
2. go ingame, try to fly in Hellfire Peninsula - flight should be disallowed. go to Nagrand, flight should be allowed

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as [WAITING TO BE TESTED].

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

REMEMBER: when testing a PR that changes something generic (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but especially check that the PR does not cause any regression (i.e. introducing new bugs).

For example: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but we should test Y and Z as well.

### CODE REVIEW

File: src/server/game/Scripting/ScriptMgr.h [hunk 1]
```diff
@@ -1449,6 +1449,16 @@ class PlayerScript : public ScriptObject
...
 
+    /**
+     * @brief This hook called before other CanFlyChecks are applied
+     *
+     * @param player Contains information about the Player
+     * @param mapId Contains information about the current map id
+     * @param zoneId Contains information about the current zone
+     * @param bySpell Contains information about the spell that invoked the check
+     */
+    [[nodiscard]] virtual bool OnCanPlayerFlyInZone(const Player* /*player*/, uint32 /*mapId*/, uint32 /*zoneId*/, SpellInfo const* /*bySpell*/) { return true; }
```
Comment (Kitzunu) [inline src/server/game/Scripting/ScriptMgr.h:1460]: Player const*

File: src/server/game/Entities/Player/Player.cpp [hunk 2]
```diff
@@ -13557,6 +13557,9 @@ uint32 Player::CalculateTalentsPoints() const
 
 bool Player::canFlyInZone(uint32 mapid, uint32 zone, SpellInfo const* bySpell) const
 {
+    if (!sScriptMgr->OnCanPlayerFlyInZone(this, mapid,zone,bySpell)) {
```
Comment (Kitzunu) [inline src/server/game/Entities/Player/Player.cpp:13560]: bracket on a new line please.

**Review by Kitzunu — CHANGES_REQUESTED** [review]

File: src/server/game/Entities/Player/Player.cpp (hunk 2 above)
Comment (Tralenor) [reply src/server/game/Entities/Player/Player.cpp]: thanks for pointing that out. should be resolved now.

File: src/server/game/Entities/Player/Player.cpp [hunk 3]
```diff
@@ -13560,6 +13560,10 @@ uint32 Player::CalculateTalentsPoints() const
 
 bool Player::canFlyInZone(uint32 mapid, uint32 zone, SpellInfo const* bySpell) const
 {
+    if (!sScriptMgr->OnCanPlayerFlyInZone(this, mapid,zone,bySpell))
+    {
+        return false;
+    }
```
Comment (Kitzunu) [inline src/server/game/Entities/Player/Player.cpp:13566]: newline under this one

**Review by Kitzunu — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
@Kitzunu thanks for reviewing this PR but please don't forget to update the PR status. Does this require additional review or is it ready to be merged? or does it need to be tested? please make it clear :)

**Comment by Tralenor** [comment]
to make the testing easier, I created a new PoC - which simply closes the sky over the Hellfire Peninsula.
you can find it here: https://github.com/Tralenor/poc-can-fly-hook-override

I also updated the testing instructions above.

**Comment by pangolp** [comment]
```
[ 20%] Building CXX object src/server/scripts/CMakeFiles/scripts.dir/Commands/PlayerCommand.cpp.o
/home/runner/work/poc-can-fly-hook-override/poc-can-fly-hook-override/modules/poc-can-fly-hook-override/src/PoCFlyHookOverride.cpp:19:108: fatal error: only virtual member functions can be marked 'override'
    bool OnCanPlayerFlyInZone(const Player* player, uint32 mapId, uint32 zoneId, SpellInfo const* bySpell) override
                                                                                                           ^~~~~~~~
1 error generated.
make[2]: *** [modules/CMakeFiles/modules.dir/build.make:90: modules/CMakeFiles/modules.dir/poc-can-fly-hook-override/src/PoCFlyHookOverride.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1199: modules/CMakeFiles/modules.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

**Comment by pangolp** [comment]
Accompanied by the module, it works fine. Here are the tests.

Here we can see that when I have the module in this case, it does not allow me to use the mount.

WoWScrnShot_081523_073213

By removing the module, it can be invoked normally.

WoWScrnShot_081523_073452

If you need more proof let me know. Anyway, the action of the module, showed an error...

```
[ 20%] Building CXX object src/server/scripts/CMakeFiles/scripts.dir/Commands/PlayerCommand.cpp.o
/home/runner/work/poc-can-fly-hook-override/poc-can-fly-hook-override/modules/poc-can-fly-hook-override/src/PoCFlyHookOverride.cpp:19:108: fatal error: only virtual member functions can be marked 'override'
    bool OnCanPlayerFlyInZone(const Player* player, uint32 mapId, uint32 zoneId, SpellInfo const* bySpell) override
                                                                                                           ^~~~~~~~
1 error generated.
make[2]: *** [modules/CMakeFiles/modules.dir/build.make:90: modules/CMakeFiles/modules.dir/poc-can-fly-hook-override/src/PoCFlyHookOverride.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:1199: modules/CMakeFiles/modules.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

I don't know if that should be corrected here or in the module, I think it has to do with the hook.
Greetings.

**Comment by Tralenor** [comment]
@pangolp thanks for the verification!

Could it be that the error you posted occurred when you compiled the core, with the module but without the hook? Otherwise I can't explain why that Error spawned.

File: src/server/game/Scripting/ScriptMgr.h [hunk 4]
```diff
@@ -1472,6 +1472,16 @@ class PlayerScript : public ScriptObject
...
 
+    /**
+     * @brief This hook called before other CanFlyChecks are applied
+     *
+     * @param player Contains information about the Player
+     * @param mapId Contains information about the current map id
+     * @param zoneId Contains information about the current zone
+     * @param bySpell Contains information about the spell that invoked the check
+     */
+    [[nodiscard]] virtual bool OnCanPlayerFlyInZone(Player const* /*player*/, uint32 /*mapId*/, uint32 /*zoneId*/, SpellInfo const* /*bySpell*/) { return true; }
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptMgr.h:1483]:
```suggestion
    [[nodiscard]] virtual bool OnCanPlayerFlyInZone(Player* /*player*/, uint32 /*mapId*/, uint32 /*zoneId*/, SpellInfo const* /*bySpell*/) { return true; }
```

Comment (Winfidonarleyan) [reply src/server/game/Scripting/ScriptMgr.h:1483]: Do not need to use a constant pointer

File: src/server/game/Scripting/ScriptMgr.h [hunk 5]
```diff
@@ -2433,6 +2443,7 @@ class ScriptMgr
     bool CanSendErrorAlreadyLooted(Player* player);
     void OnAfterCreatureLoot(Player* player);
     void OnAfterCreatureLootMoney(Player* player);
+    bool OnCanPlayerFlyInZone(Player const* player, uint32 mapId, uint32 zoneId, SpellInfo const* bySpell);
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptMgr.h:2446]:
```suggestion
    bool OnCanPlayerFlyInZone(Player* player, uint32 mapId, uint32 zoneId, SpellInfo const* bySpell);
```

File: src/server/game/Scripting/ScriptDefines/PlayerScript.cpp [hunk 6]
```diff
@@ -74,6 +74,21 @@ void ScriptMgr::OnPlayerReleasedGhost(Player* player)
     });
 }
 
+bool ScriptMgr::OnCanPlayerFlyInZone(Player const* player, uint32 mapId, uint32 zoneId, SpellInfo const* bySpell)
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptDefines/PlayerScript.cpp:77]:
```suggestion
bool ScriptMgr::OnCanPlayerFlyInZone(Player* player, uint32 mapId, uint32 zoneId, SpellInfo const* bySpell)
```

File: src/server/game/Scripting/ScriptDefines/PlayerScript.cpp [hunk 7]
```diff
@@ -74,6 +74,21 @@ void ScriptMgr::OnPlayerReleasedGhost(Player* player)
...
+{
+    auto ret = IsValidBoolScript<PlayerScript>([&](PlayerScript* script)
```
Comment (Winfidonarleyan) [inline src/server/game/Scripting/ScriptDefines/PlayerScript.cpp:79]:
```suggestion
    auto ret = IsValidBoolScript<PlayerScript>([player, mapId, zoneId, bySpell](PlayerScript* script)
```

File: src/server/game/Scripting/ScriptMgr.h (hunk 4 above)
Comment (Tralenor) [reply src/server/game/Scripting/ScriptMgr.h:1483]: I have nothing against it, but what is the benefit of a non const pointer over a const pointer? In which usecase would you anticipate that the additional flexibility becomes useful?

Comment (Winfidonarleyan) [reply src/server/game/Scripting/ScriptMgr.h]: For scripts and other better `Player*`. All functions can be used

Comment (Tralenor) [reply src/server/game/Scripting/ScriptMgr.h]: Ah I see, sounds good. Well at least in theory. These changes do not successfully compile (at least not without further changes)
The reason is that the function is called like this:
 
```
`bool Player::canFlyInZone(uint32 mapid, uint32 zone, SpellInfo const* bySpell) const ` like: if (!sScriptMgr->OnCanPlayerFlyInZone(this, mapid,zone,bySpell))
    {
        return false;
    }
    
```
the object **this** already references a const * player. Which in turn stems from the fact, its called from within:
```
SpellCastResult SpellInfo::CheckLocation(uint32 map_id, uint32 zone_id, uint32 area_id, Player const* player /*= nullptr*/, bool strict /*= true*/) const
{....}
```
Essentially, if we want to provide this hook without the const restriction we need to loosen that restriction earlier in the call stack as well. That is not too hard to do, but the question is whether or not we really want it.

Comment (Tralenor) [reply src/server/game/Scripting/ScriptMgr.h]: I adjusted the code accordingly, now we don't have const * player anymore

File: src/server/game/Spells/SpellInfo.cpp [hunk 8]
```diff
@@ -1470,7 +1470,7 @@ SpellCastResult SpellInfo::CheckShapeshift(uint32 form) const
     return SPELL_CAST_OK;
 }
 
-SpellCastResult SpellInfo::CheckLocation(uint32 map_id, uint32 zone_id, uint32 area_id, Player const* player /*= nullptr*/, bool strict /*= true*/) const
+SpellCastResult SpellInfo::CheckLocation(uint32 map_id, uint32 zone_id, uint32 area_id, Player * player /*= nullptr*/, bool strict /*= true*/) const
```
Comment (Kitzunu) [inline src/server/game/Spells/SpellInfo.cpp:1473]:
```suggestion
SpellCastResult SpellInfo::CheckLocation(uint32 map_id, uint32 zone_id, uint32 area_id, Player* player /*= nullptr*/, bool strict /*= true*/) const
```

File: src/server/game/Spells/SpellInfo.h [hunk 9]
```diff
@@ -474,7 +474,7 @@ friend class SpellMgr;
     bool IsAuraExclusiveBySpecificPerCasterWith(SpellInfo const* spellInfo) const;
 
     SpellCastResult CheckShapeshift(uint32 form) const;
-    SpellCastResult CheckLocation(uint32 map_id, uint32 zone_id, uint32 area_id, Player const* player = nullptr, bool strict = true) const;
+    SpellCastResult CheckLocation(uint32 map_id, uint32 zone_id, uint32 area_id, Player * player = nullptr, bool strict = true) const;
```
Comment (Kitzunu) [inline src/server/game/Spells/SpellInfo.h:477]:
```suggestion
    SpellCastResult CheckLocation(uint32 map_id, uint32 zone_id, uint32 area_id, Player* player = nullptr, bool strict = true) const;
```

**Review by Kitzunu — APPROVED** [review]

**Pull request merged**